### ISSUE TITLE
Make waiting for event loop implicit or explicit in all cases

### ISSUE THREAD

**Issue by jliuhtonen** [assaf/zombie#1054]
Zombie.js does not start to process its event queue until you call `browser.wait()`. 

In some `browser` methods, like `click()`, this call to wait is implicit, meaning the method calls `wait()` for you or returns the lazy promise returned by `browser.fire()`. Then again some browser methods, like `fill()` ignore the lazy promise returned by `browser.fire()` and return `this`. In the latter case you must call `browser.wait()` in your test code explicitly. This is unexpected and not well documented and leads to confusing test code.

I personally would prefer an implicit approach, so all browser methods emitting events would implicitly run `wait()` by returning the promise returned by `browser.fire()`, but either way, all the methods should have the same behaviour.

**Comment by assaf** [comment]
If you pass a callback to a method that takes it, it will call `wait` for you; if you wait on the promise a method returns, it will `wait` for you. So the behavior is very consistent, but only some methods take a callback/return a promise. You don't have to use this feature, you can treat all methods as not taking promise/returning callbacks, and call `wait` when necessary.

**Comment by jliuhtonen** [comment]
Why does `click()` accept a callback and/or return a promise and wait, but for example `fill()` and `selectOption()` do not?

If all methods worked in the same way, then you wouldn't have to remember which one waits and which one does not, although they all generate events. Making the API work the same way for all methods that create events would IMO improve the usability of the API and make the code calling Zombie cleaner.

Consider the following example:

```
it('adds an item to shopping basket', () => {
  return browser.click('.product-catalog')
    .then(() => {
      browser.fill('.search', 'macbook pro')
      return browser.wait() // this is here because fill does not return a promise or wait like click() does
    })
    .tap(() => browser.assert.elements('.product', 1, 'should find just one product'))
    .then(() => browser.click('.add-to-basket'))
    .tap(() => browser.assert.elements('.basket-item', 1))
})
```

This would become:
```
it('adds an item to shopping basket', () => {
  return browser.click('.product-catalog')
    .then(() => browser.fill('.search', 'macbook pro'))
    .tap(() => browser.assert.elements('.product', 1, 'should find just one product'))
    .then(() => browser.click('.add-to-basket'))
    .tap(() => browser.assert.elements('.basket-item', 1))
})
```

**Comment by jliuhtonen** [comment]
Thanks for the clarification. That is a good shortcut. Though I would argue that many web apps have features involving events, like live search as you type or client side form validations, that would make the shortcut worth implementing to the whole browser API. The API would then also resemble more closely using a browser with GUI. :)

**Comment by assaf** [comment]
Yes, definitely some people will find it more useful if `fill` also did a `wait`, just like some people find it more useful that you can chain multiple `fill`. Neither of these shortcuts will please everyone, I just find the current API to be the most useful.

You asked about making `wait` implicit or explicit for all methods. Explicit is already supported: call the method without a callback, then follow up with a call to `wait`.

**Comment by jliuhtonen** [comment]
Fair enough.

I was referring more to the easiness of constructing a mental model of how Zombie works and how to write (and read) code that triggers DOM events inside it. IMO it would make things easier if there no special cases and exceptions for doing similar things (interacting with the page DOM).

Sure, one can implement a utility that does the waiting or call `.wait()` always explicitly, but that's beside the point.

Btw. the way I discovered this was that I tried to implement a waiting assertion that checks periodically whether a given predicate is true or times out with an assertion error after a given period. Using one of the methods that do not wait of course I ended up always timing out, because the event loop was not run at all. After this I read the source and learned that `.wait()` does more than waiting. I now understand why it is, but it's still a surprising implementation detail that the event loop does not run by itself and you have to call a method called `.wait()` (sometimes) to make it run.

The reason I brought this up is that I really like Zombie's approach to headless browser testing with Node and I hope that it can be made even better in the future. Thanks for your work! 👍
